File: sdks/typescript/src/brain.ts
/**
 * High-level Brain facade — the canonical entry point for the
 * TypeScript SDK.  Mirrors the Python `thinkingroot.Brain` shape so
 * polyglot teams move between languages without re-learning the API.
 *
 * Three constructors:
 *   - `Brain.remote(url)`  — explicit URL.
 *   - `Brain.connect()`    — cortex-aware auto-discovery.
 *   - `Brain.mount(path)`  — spawn `root mount`, attach to result.
 *
 * Pure-TS / pure-fetch — no native bindings, no runtime deps beyond
 * Node 18's built-in `fetch`.
 *
 * Spec: `docs/secondary-brain-concept.md` §4.
 */

import { spawn } from "node:child_process";
import { randomBytes } from "node:crypto";

import { Client, type ClientOptions } from "./client.js";
import { ApiError, ConnectionError } from "./errors.js";
import {
  DEFAULT_HOST,
  DEFAULT_PORT,
  processAlive,
  readLock,
} from "./cortex.js";
import type {
  Claim,
  Entity,
  EngramRef,
  EngramScope,
  HybridResponse,
  MaterializeResponse,
  MountSummary,
  ProbeAnswer,
  RetrievalRequest,
  SearchResult,
  WorkspaceInfo,
} from "./types.js";

export interface BrainOptions {
  /** Workspace name.  Defaults to the first workspace mounted on the daemon. */
  workspace?: string;
  /** Bring-your-own session id.  Defaults to a fresh `ts-<16 hex>` token. */
  sessionId?: string;
  /** API token for password-protected daemons. */
  apiKey?: string | null;
  /** Per-request timeout in milliseconds.  Default 120s. */
  timeoutMs?: number;
  /** Override the global `fetch` (testing seam). */
  fetch?: typeof fetch;
}

export interface BrainInfo {
  transport: "remote";
  workspace: string;
  baseUrl: string;
  sessionId: string;
  daemonPid: number | null;
  daemonStartedBy: string | null;
}

const SESSION_HEADER = "X-TR-Session-Id";

/**
 * Unified facade for talking to a ThinkingRoot daemon.
 *
 * Construct via the `remote`, `connect`, or `mount` static methods —
 * never the constructor directly.
 */
export class Brain {
  private readonly _client: Client;
  private readonly _workspace: string;
  private readonly _sessionId: string;
  private readonly _baseUrl: string;
  private _daemonPid: number | null = null;
  private _daemonStartedBy: string | null = null;

  private constructor(
    client: Client,
    workspace: string,
    sessionId: string,
    baseUrl: string,
  ) {
    this._client = client;
    this._workspace = workspace;
    this._sessionId = sessionId;
    this._baseUrl = baseUrl;
  }

  // ─── Constructors ──────────────────────────────────────────

  /**
   * Attach to a running daemon at `baseUrl`.  When `workspace` is
   * omitted, the first workspace mounted on the daemon is used.
   */
  static async remote(
    baseUrl: string = `http://${DEFAULT_HOST}:${DEFAULT_PORT}`,
    opts: BrainOptions = {},
  ): Promise<Brain> {
    const clientOpts: ClientOptions = {
      baseUrl,
      apiKey: opts.apiKey,
      timeoutMs: opts.timeoutMs,
      fetch: opts.fetch,
    };
    const client = new Client(clientOpts);
    const workspace = opts.workspace ?? (await resolveFirstWorkspace(client));
    const sessionId = opts.sessionId ?? newSessionId();
    return new Brain(client, workspace, sessionId, client.baseUrl);
  }

  /**
   * Cortex-aware auto-discovery.  Reads the cortex lockfile and
   * attaches to the daemon when one is alive.  Throws
   * {@link ConnectionError} otherwise — the pure-fetch SDK has no
   * in-process fallback (that's what the Rust + Python SDKs offer).
   */
  static async connect(opts: BrainOptions = {}): Promise<Brain> {
    const lock = await readLock();
    if (lock === null) {
      throw new ConnectionError(
        "no cortex daemon running. Start one with `root serve` and " +
          "retry, or use Brain.remote(url) with an explicit URL.",
      );
    }
    if (!processAlive(lock.pid)) {
      throw new ConnectionError(
        `cortex.lock points to pid ${lock.pid} but the process is dead. ` +
          "Run `root serve` to start a fresh daemon.",
      );
    }
    const baseUrl = `http://${lock.host}:${lock.port}`;
    const brain = await Brain.remote(baseUrl, opts);
    brain._daemonPid = lock.pid;
    brain._daemonStartedBy = lock.started_by;
    return brain;
  }

  /**
   * Mount a `.tr` pack via the `root mount` CLI subcommand and
   * return a Brain attached to the freshly-mounted workspace.
   *
   * Requires the `root` binary on `$PATH`.
   */
  static async mount(
    packPath: string,
    options: { name?: string; noVerify?: boolean; recompile?: boolean } = {},
  ): Promise<Brain> {
    const args = ["mount", packPath];
    if (options.name) args.push("--name", options.name);
    if (options.noVerify) args.push("--no-verify");
    if (options.recompile) args.push("--recompile");

    const summary = await runRootCli<MountSummary>(args);
    const restUrl = summary.rest_url.replace(/\/$/, "");
    const marker = "/api/v1/ws/";
    const baseUrl = restUrl.includes(marker)
      ? restUrl.split(marker)[0]!
      : restUrl;

    const brain = await Brain.remote(baseUrl, {
      workspace: summary.workspace,
    });
    brain._daemonPid = summary.daemon_pid;
    brain._daemonStartedBy = "root_mount";
    return brain;
  }

  // ─── Introspection ─────────────────────────────────────────

  get workspace(): string {
    return this._workspace;
  }

  get sessionId(): string {
    return this._sessionId;
  }

  info(): BrainInfo {
    return {
      transport: "remote",
      workspace: this._workspace,
      baseUrl: this._baseUrl,
      sessionId: this._sessionId,
      daemonPid: this._daemonPid,
      daemonStartedBy: this._daemonStartedBy,
    };
  }

  // ─── Workspace listing ─────────────────────────────────────

  workspaces(): Promise<WorkspaceInfo[]> {
    return this._client.get<WorkspaceInfo[]>("/workspaces");
  }

  // ─── Claims / Entities / Search ────────────────────────────

  entities(): Promise<Entity[]> {
    return this._client.get<Entity[]>(`/ws/${this._workspace}/entities`);
  }

  entity(name: string): Promise<Entity> {
    return this._client.get<Entity>(
      `/ws/${this._workspace}/entities/${encodeURIComponent(name)}`,
    );
  }

  claims(opts: {
    type?: string;
    minConfidence?: number;
    limit?: number;
  } = {}): Promise<Claim[]> {
    const params = new URLSearchParams();
    if (opts.type) params.set("type", opts.type);
    if (opts.minConfidence !== undefined)
      params.set("min_confidence", String(opts.minConfidence));
    if (opts.limit !== undefined) params.set("limit", String(opts.limit));
    const qs = params.toString();
    const path = `/ws/${this._workspace}/claims${qs ? `?${qs}` : ""}`;
    return this._client.get<Claim[]>(path);
  }

  search(query: string, topK: number = 10): Promise<SearchResult> {
    const params = new URLSearchParams({ q: query, top_k: String(topK) });
    return this._client.get<SearchResult>(
      `/ws/${this._workspace}/search?${params.toString()}`,
    );
  }

  // ─── Hybrid Retrieval ──────────────────────────────────────

  hybridSearch(
    query: string,
    opts: Partial<RetrievalRequest> = {},
  ): Promise<HybridResponse> {
    const body: RetrievalRequest = {
      query_text: query,
      session_id: this._sessionId,
      typed_predicates: [],
      clearance: ["public"],
      top_k: 20,
      time_window: null,
      scoring_profile: "default",
      require_certificate: false,
      include_test_origin: true,
      include_quarantined: false,
      require_provenance_verified: false,
      now: null,
      scoped_claim_ids: null,
      ...opts,
    };
    return this._client.post<HybridResponse>(
      `/ws/${this._workspace}/search/hybrid`,
      body,
    );
  }

  // ─── RARP / Active Engram Protocol ─────────────────────────

  materializeEngram(
    topic: string,
    opts: { seedEntityIds?: string[]; scope?: EngramScope } = {},
  ): Promise<MaterializeResponse> {
    const body: Record<string, unknown> = { topic };
    if (opts.seedEntityIds) body["seed_entity_ids"] = opts.seedEntityIds;
    if (opts.scope) body["scope"] = opts.scope;
    return this._client.post<MaterializeResponse>(
      `/ws/${this._workspace}/engrams`,
      body,
      { [SESSION_HEADER]: this._sessionId },
    );
  }

  probe(
    pointer: string,
    question: string,
    opts: {
      clearance?: string[];
      probeKind?: string;
      scoreWithHybrid?: boolean;
    } = {},
  ): Promise<ProbeAnswer> {
    const body: Record<string, unknown> = {
      question,
      score_with_hybrid: opts.scoreWithHybrid ?? false,
    };
    if (opts.clearance) body["clearance"] = opts.clearance;
    if (opts.probeKind) body["probe_kind"] = opts.probeKind;
    return this._client.post<ProbeAnswer>(
      `/ws/${this._workspace}/engrams/${encodeURIComponent(pointer)}/probe`,
      body,
      { [SESSION_HEADER]: this._sessionId },
    );
  }

  engrams(): Promise<EngramRef[]> {
    return this._client.get<EngramRef[]>(`/ws/${this._workspace}/engrams`, {
      [SESSION_HEADER]: this._sessionId,
    });
  }

  async expire(pointer: string): Promise<boolean> {
    const result = await this._client.del<{ expired: boolean; pointer: string }>(
      `/ws/${this._workspace}/engrams/${encodeURIComponent(pointer)}`,
      { [SESSION_HEADER]: this._sessionId },
    );
    return result.expired;
  }

  /**
   * Drop every engram in this Brain's session.  Idempotent.
   */
  async resetSession(): Promise<void> {
    const refs = await this.engrams();
    for (const ref of refs) {
      try {
        await this.expire(ref.pointer);
      } catch (err) {
        if (err instanceof ApiError && err.statusCode === 404) {
          // Already gone — fine.
          continue;
        }
        throw err;
      }
    }
  }
}

// ─── Module-private helpers ──────────────────────────────────

async function resolveFirstWorkspace(client: Client): Promise<string> {
  const workspaces = await client.get<WorkspaceInfo[]>("/workspaces");
  if (workspaces.length === 0) {
    throw new ApiError(
      404,
      "NO_WORKSPACE",
      "No workspaces mounted on the daemon",
    );
  }
  return workspaces[0]!.name;
}

function newSessionId(): string {
  return `ts-${randomBytes(8).toString("hex")}`;
}

/**
 * Spawn `root <args>`, capture stdout, parse JSON.  Used by
 * {@link Brain.mount}.
 */
function runRootCli<T>(args: readonly string[]): Promise<T> {
  return new Promise((resolve, reject) => {
    const child = spawn("root", args, { stdio: ["ignore", "pipe", "pipe"] });
    const stdoutChunks: Buffer[] = [];
    const stderrChunks: Buffer[] = [];

    child.stdout.on("data", (chunk: Buffer) => stdoutChunks.push(chunk));
    child.stderr.on("data", (chunk: Buffer) => stderrChunks.push(chunk));

    child.on("error", (err) => {
      if ((err as NodeJS.ErrnoException).code === "ENOENT") {
        reject(
          new ConnectionError(
            "`root` binary not on $PATH. Install via `cargo install thinkingroot-cli`.",
            err,
          ),
        );
      } else {
        reject(new ConnectionError(`spawn root: ${err.message}`, err));
      }
    });

    child.on("close", (code) => {
      if (code !== 0) {
        const stderr = Buffer.concat(stderrChunks).toString("utf-8");
        reject(
          new ConnectionError(
            `root ${args.join(" ")} exited ${code}: ${stderr}`,
          ),
        );
        return;
      }
      const stdout = Buffer.concat(stdoutChunks).toString("utf-8");
      try {
        resolve(JSON.parse(stdout) as T);
      } catch (err) {
        reject(
          new ConnectionError(
            `root ${args.join(" ")} produced non-JSON output: ${stdout}`,
            err,
          ),
        );
      }
    });
  });
}

File: thinkingroot-python/python/thinkingroot/brain.py
"""High-level facade for ThinkingRoot in Python.

The :class:`Brain` class is the canonical user-facing entry point of
the SDK.  It abstracts over three transports:

============== ===========================================================
``open(path)``  In-process via PyO3.  Zero-network, sub-millisecond
                queries.  Best for notebook / data-science usage.
``remote(url)`` HTTP via httpx.  Best for distributed / containerized
                agents that talk to a central daemon.
``connect()``   Cortex-aware auto-discovery.  Reads the cortex lockfile
                — if a daemon is running, attaches; otherwise falls
                back to in-process.  Best for the "just works"
                developer experience.
============== ===========================================================

All three transports expose the same method surface (entities,
claims, search, hybrid_search, materialize_engram, probe, engrams,
expire) — so swapping transports is a one-line change.

Example::

    from thinkingroot import Brain

    brain = Brain.connect()
    pointer = brain.materialize_engram("auth flow")["pointer"]
    answer = brain.probe(pointer, "what changed last week?")
    for row, claim_id in zip(answer["answer"], answer["claim_ids"]):
        print(claim_id, row)

Spec: ``docs/secondary-brain-concept.md`` §4 (the SDK plug story).
"""

from __future__ import annotations

import secrets
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from thinkingroot import cortex
from thinkingroot.client import APIError, Client


@dataclass(frozen=True)
class BrainInfo:
    """Connection metadata returned by :meth:`Brain.info`."""

    transport: str
    workspace: str
    base_url: str | None
    session_id: str
    daemon_pid: int | None
    daemon_started_by: str | None


class Brain:
    """Unified facade over in-process and remote ThinkingRoot engines.

    Construct via the :meth:`open`, :meth:`remote`, :meth:`mount`, or
    :meth:`connect` class methods — never the ``__init__`` directly.
    """

    def __init__(
        self,
        *,
        transport: str,
        engine: Any | None,
        client: Client | None,
        workspace: str,
        session_id: str,
        base_url: str | None,
        daemon_pid: int | None = None,
        daemon_started_by: str | None = None,
    ):
        self._transport = transport
        self._engine = engine
        self._client = client
        self._workspace = workspace
        self._session_id = session_id
        self._base_url = base_url
        self._daemon_pid = daemon_pid
        self._daemon_started_by = daemon_started_by

    # ─── Constructors ──────────────────────────────────────────

    @classmethod
    def open(cls, path: str | Path) -> "Brain":
        """Open a compiled workspace in-process.

        The directory must already contain ``.thinkingroot/`` (i.e.
        a workspace produced by ``root compile`` or
        ``thinkingroot.compile()``).  For a ``.tr`` pack, use
        :meth:`mount` instead.
        """
        # Lazy import: native bindings may not be present in pure-HTTP
        # installs (e.g. wheels built with --no-default-features).
        from thinkingroot import open as _native_open

        engine = _native_open(str(path))
        # Workspace name = directory name, matching the Rust open()
        # behaviour.  The session id rides on the PyO3 Engine.
        ws_name = Path(path).resolve().name or "default"
        return cls(
            transport="in_process",
            engine=engine,
            client=None,
            workspace=ws_name,
            session_id=engine.session_id,
            base_url=None,
        )

    @classmethod
    def remote(
        cls,
        base_url: str = "http://127.0.0.1:31760",
        *,
        api_key: str | None = None,
        workspace: str | None = None,
        session_id: str | None = None,
    ) -> "Brain":
        """Attach to a running ``root serve`` daemon over HTTP.

        ``workspace`` defaults to the first workspace mounted on the
        daemon.  ``session_id`` defaults to a fresh per-Brain token —
        passing a stable value across processes lets multiple Brains
        share the same engram set.
        """
        client = Client(base_url=base_url, api_key=api_key)
        ws_name = workspace or _resolve_first_workspace(client)
        sid = session_id or _new_session_id()
        return cls(
            transport="remote",
            engine=None,
            client=client,
            workspace=ws_name,
            session_id=sid,
            base_url=base_url,
        )

    @classmethod
    def connect(
        cls,
        *,
        workspace: str | None = None,
        session_id: str | None = None,
    ) -> "Brain":
        """Cortex-aware auto-discovery.

        Reads the cortex lockfile.  If a daemon is alive at the
        recorded host:port, returns a :meth:`remote` Brain attached
        to it.  Otherwise raises :class:`ConnectionError` — no
        in-process fallback because that would defeat the
        single-writer guarantee cortex enforces (a Brain.connect()
        opening CozoDB while the daemon also has it open is the
        write-conflict bug cortex was built to prevent).
        """
        lock = cortex.read_lock()
        if lock is None:
            raise ConnectionError(
                "no cortex daemon running. Start one with `root serve` "
                "or use Brain.open(path) for in-process access."
            )
        if not cortex.process_alive(lock.pid):
            raise ConnectionError(
                f"cortex.lock points to pid {lock.pid} but the process is dead. "
                "Run `root serve` to start a fresh daemon."
            )
        return cls.remote(
            base_url=lock.base_url,
            workspace=workspace,
            session_id=session_id,
        ).with_daemon_meta(pid=lock.pid, started_by=lock.started_by)

    @classmethod
    def mount(
        cls,
        pack_path: str | Path,
        *,
        name: str | None = None,
        no_verify: bool = False,
        recompile: bool = False,
    ) -> "Brain":
        """Mount a ``.tr`` pack via the cortex daemon and attach.

        Convenience wrapper around the ``root mount`` CLI subcommand —
        spawns it as a subprocess, parses the printed ``MountSummary``
        JSON, then returns a :meth:`remote` Brain pointed at the
        freshly-mounted workspace.

        Requires the ``root`` binary on ``PATH``.  For pure in-process
        unpack-and-replay without a daemon, use the lower-level Rust
        ``mount_cmd`` — the daemon path is the supported "secondary
        brain plug" experience.
        """
        import json
        import shutil
        import subprocess

        binary = shutil.which("root")
        if binary is None:
            raise RuntimeError(
                "`root` binary not on PATH. Install via `cargo install "
                "thinkingroot-cli` or download a release artifact."
            )
        cmd = [binary, "mount", str(pack_path)]
        if name:
            cmd.extend(["--name", name])
        if no_verify:
            cmd.append("--no-verify")
        if recompile:
            cmd.append("--recompile")
        result = subprocess.run(
            cmd, capture_output=True, text=True, check=False
        )
        if result.returncode != 0:
            raise RuntimeError(
                f"root mount failed (exit {result.returncode}): {result.stderr}"
            )
        try:
            summary = json.loads(result.stdout)
        except json.JSONDecodeError as exc:
            raise RuntimeError(
                f"root mount returned non-JSON output: {result.stdout!r}"
            ) from exc
        # rest_url shape: http://host:port/api/v1/ws/<name>/.  Strip
        # the /api/v1/ws/<name>/ tail so the Brain.remote(base_url)
        # gets the daemon root.
        rest = summary["rest_url"].rstrip("/")
        marker = "/api/v1/ws/"
        if marker in rest:
            base = rest.split(marker, 1)[0]
        else:
            base = rest
        return cls.remote(
            base_url=base,
            workspace=summary["workspace"],
        ).with_daemon_meta(
            pid=int(summary.get("daemon_pid", 0)),
            started_by="root_mount",
        )

    def with_daemon_meta(
        self,
        *,
        pid: int | None,
        started_by: str | None,
    ) -> "Brain":
        """Internal: stamp daemon provenance on a Brain."""
        self._daemon_pid = pid
        self._daemon_started_by = started_by
        return self

    # ─── Introspection ─────────────────────────────────────────

    @property
    def workspace(self) -> str:
        return self._workspace

    @property
    def session_id(self) -> str:
        return self._session_id

    @property
    def transport(self) -> str:
        return self._transport

    def info(self) -> BrainInfo:
        return BrainInfo(
            transport=self._transport,
            workspace=self._workspace,
            base_url=self._base_url,
            session_id=self._session_id,
            daemon_pid=self._daemon_pid,
            daemon_started_by=self._daemon_started_by,
        )

    # ─── Claims / Entities / Search ────────────────────────────

    def entities(self) -> list[dict]:
        if self._engine is not None:
            return self._engine.get_entities()
        return self._client.entities(workspace=self._workspace)

    def entity(self, name: str) -> dict:
        if self._engine is not None:
            return self._engine.get_entity(name)
        return self._client.entity(name, workspace=self._workspace)

    def claims(
        self,
        *,
        claim_type: str | None = None,
        min_confidence: float | None = None,
    ) -> list[dict]:
        if self._engine is not None:
            return self._engine.get_claims(
                **{"type": claim_type} if claim_type else {},
                min_confidence=min_confidence,
            )
        return self._client.claims(
            workspace=self._workspace,
            type=claim_type,
            min_confidence=min_confidence,
        )

    def relations(self, entity: str) -> list[dict]:
        if self._engine is not None:
            return self._engine.get_relations(entity)
        return self._client.relations(entity, workspace=self._workspace)

    def search(self, query: str, *, top_k: int = 10) -> dict:
        if self._engine is not None:
            return self._engine.search(query, top_k=top_k)
        return self._client.search(query, workspace=self._workspace, top_k=top_k)

    # ─── Hybrid Retrieval ──────────────────────────────────────

    def hybrid_search(
        self,
        query: str,
        *,
        top_k: int = 20,
        require_certificate: bool = False,
        include_quarantined: bool = False,
        require_provenance_verified: bool = False,
    ) -> dict:
        """Run the 11-component hybrid retrieval pipeline.

        Returns the full ``HybridResponse`` shape: a list of
        ``hits`` (each carrying ``claim_id``, ``score``,
        ``score_breakdown``, ``provenance``) plus diagnostic
        ``stage_timings`` and ``routing`` shape.
        """
        if self._engine is not None:
            return self._engine.hybrid_search(
                query,
                top_k=top_k,
                require_certificate=require_certificate,
                include_quarantined=include_quarantined,
                require_provenance_verified=require_provenance_verified,
            )
        # Remote: POST /api/v1/ws/{ws}/search/hybrid with the full
        # RetrievalRequest body shape (matches engine::RetrievalRequest
        # so the SDK and the engine speak the same wire schema).
        body = {
            "query_text": query,
            "typed_predicates": [],
            "session_id": self._session_id,
            "clearance": ["public"],
            "top_k": top_k,
            "time_window": None,
            "scoring_profile": "default",
            "require_certificate": require_certificate,
            "include_test_origin": True,
            "include_quarantined": include_quarantined,
            "require_provenance_verified": require_provenance_verified,
            "now": None,
            "scoped_claim_ids": None,
        }
        return self._client._post_json(
            f"/ws/{self._workspace}/search/hybrid", body
        )

    # ─── RARP / Active Engram Protocol ─────────────────────────

    def materialize_engram(
        self,
        topic: str,
        *,
        seed_entity_ids: list[str] | None = None,
        scope: dict | None = None,
    ) -> dict:
        """Build an Engram for ``topic``. Returns ``{pointer, summary}``."""
        if self._engine is not None:
            return self._engine.materialize_engram(
                topic, seed_entity_ids=seed_entity_ids, scope=scope
            )
        body = {"topic": topic}
        if seed_entity_ids is not None:
            body["seed_entity_ids"] = seed_entity_ids
        if scope is not None:
            body["scope"] = scope
        return self._client_post_with_session(
            f"/ws/{self._workspace}/engrams", body
        )

    def probe(
        self,
        pointer: str,
        question: str,
        *,
        clearance: list[str] | None = None,
        probe_kind: str | None = None,
        score_with_hybrid: bool = False,
    ) -> dict:
        """Probe an Engram with a natural-language question.

        Returns the ``ProbeAnswer`` shape (``answer`` rows + parallel
        ``claim_ids``, ``source_byte_spans``, ``source_authority``,
        ``source_blake3s`` arrays + trust/temporal/lineage/privacy
        fields + ``caveats``).
        """
        if self._engine is not None:
            return self._engine.probe_engram(
                pointer,
                question,
                clearance=clearance,
                probe_kind=probe_kind,
                score_with_hybrid=score_with_hybrid,
            )
        body = {"question": question, "score_with_hybrid": score_with_hybrid}
        if clearance is not None:
            body["clearance"] = clearance
        if probe_kind is not None:
            body["probe_kind"] = probe_kind
        return self._client_post_with_session(
            f"/ws/{self._workspace}/engrams/{pointer}/probe", body
        )

    def engrams(self) -> list[dict]:
        """List active engrams for this Brain's session."""
        if self._engine is not None:
            return self._engine.list_engrams()
        return self._client_get_with_session(
            f"/ws/{self._workspace}/engrams"
        )

    def expire(self, pointer: str) -> bool:
        """Drop an engram. Returns True when one was removed."""
        if self._engine is not None:
            return self._engine.expire_engram(pointer)
        result = self._client_delete_with_session(
            f"/ws/{self._workspace}/engrams/{pointer}"
        )
        return bool(result.get("expired", False))

    def reset_session(self) -> None:
        """Drop every engram in this Brain's session."""
        if self._engine is not None:
            self._engine.reset_session()
            return
        # Remote: there's no explicit /sessions/{id} DELETE endpoint
        # (engrams are per-session, so dropping each one accomplishes
        # the same). We list-and-evict; idempotent.
        for ref in self.engrams():
            self.expire(ref["pointer"])

    # ─── Internal HTTP helpers (session-aware) ─────────────────
    #
    # The bare `Client` doesn't know about session ids — engram
    # endpoints expect the X-TR-Session-Id header.  These helpers
    # inject it without forking the existing Client public API.

    def _client_post_with_session(self, path: str, body: dict) -> Any:
        assert self._client is not None
        resp = self._client._client.post(
            f"{self._client._base}{path}",
            json=body,
            headers={"X-TR-Session-Id": self._session_id},
        )
        return self._client._handle(resp)

    def _client_get_with_session(self, path: str) -> Any:
        assert self._client is not None
        resp = self._client._client.get(
            f"{self._client._base}{path}",
            headers={"X-TR-Session-Id": self._session_id},
        )
        return self._client._handle(resp)

    def _client_delete_with_session(self, path: str) -> Any:
        assert self._client is not None
        resp = self._client._client.delete(
            f"{self._client._base}{path}",
            headers={"X-TR-Session-Id": self._session_id},
        )
        return self._client._handle(resp)


def _resolve_first_workspace(client: Client) -> str:
    """Return the name of the first workspace mounted on ``client``.

    Raises :class:`APIError` (with ``code='NO_WORKSPACE'``) if none
    are mounted — same shape the bare ``Client._resolve_workspace``
    raises so callers don't need to handle two error styles.
    """
    workspaces = client.workspaces()
    if not workspaces:
        raise APIError(
            status_code=404,
            code="NO_WORKSPACE",
            message="No workspaces mounted on the daemon",
        )
    return workspaces[0]["name"]


def _new_session_id() -> str:
    """Mint a fresh Brain session id.  Matches the Rust pattern of
    ``py-<16 hex>`` for cross-language traceability."""
    return f"py-{secrets.token_hex(8)}"
